Add CheckDeps, InitDeps, CloseDeps helper

edwarnicke opened the pull request:
The Deps pattern in Plugins is a powerful way to allow
users to wire in alternate implementations for things the Plugin
uses. In addition to the flexibility it offers, it also makes
it much much easier to Mock and test.

Unfortunately, there are several mistakes its easy to make with Deps:

- Making a Deps field private (lowercase) which precludes wiring in
  an alternate implementation
- Failing to set a Default value for Deps in options.go

CheckDeps checks for all of these mistakes. It is suitable to
be run in the NewPlugin(...) in options.go as a final check before
returning a new Plugin to insure its ready to Go :)

*****************

InitDeps - Runs Init() on all Deps implemented with Plugins

Deps should be interfaces, they do not have to be Plugins
InitDeps loops through all of the fields in Deps, and if they
are Plugins, it Init()'s them.


*****************

CloseDeps - Runs Close() on all Deps implemented with Plugins

Closes all Deps of the given plugin if they themselves
are idempotent.Plugins

edwarnicke commented:
Note to reviewer: #166 should be merged first as this is based on it.

fkautz commented on utils/helper/check_deps.go:
```diff
@@ -0,0 +1,99 @@
+
+func getDeps(p idempotent.PluginAPI) (*reflect.StructField, *reflect.Value, error) {
+	if p == nil {
+		return nil, nil, fmt.Errorf("Cannot call CheckDeps on nil")
```
Cannot should be lower case.

fkautz commented on utils/helper/check_deps.go:
```diff
@@ -0,0 +1,99 @@
+			field := v.Interface()
+			zeroValue := reflect.Zero(v.Type())
+			if field == nil || field == zeroValue {
+				return fmt.Errorf("field in % struct  named %s is zero valued, and lacks tag `optional:\"true\"`.  Please either set it to a non-zero value or add the `optional:\"true\"` to indicate its acceptable for it to be zero valued", structField.Name, sf.Name)
```
Looks like the first % is missing its parameter. E.g. %s, %d.

fkautz commented on utils/helper/check_deps_test.go:
```diff
@@ -0,0 +1,141 @@
+	Running() bool
+}
+
+type PluginAPI interface {
```
This exists in a test file and is unused. Should we remove it?

fkautz commented on utils/helper/check_deps_test.go:
```diff
@@ -0,0 +1,141 @@
+	Expect(helper.CheckDeps(p)).ToNot(Succeed())
+}
+
+type PluginNonInterfaceDeps struct {
```
This type is not currently used and exists in a test. Should we remove?

fkautz commented on utils/helper/close_deps_test.go:
```diff
@@ -0,0 +1,79 @@
+	Expect(p.Deps.One.Running()).To(BeFalse())
+}
+
+type PluginErrorOnClose struct {
```
This is not being used.

fkautz commented on utils/helper/init_deps.go:
```diff
@@ -0,0 +1,55 @@
+	return initDeps(sf, vdeps)
+}
+
+func initDeps(structField *reflect.StructField, value *reflect.Value) error {
```
structField is not being used. I assume this is by design. Can we rename this variable to _ so that it is clear it is unused?

fkautz commented:
Considering the size of check_deps, would it make sense to make a checkdeps package within helpers? If there is only going to be one helper, perhaps we can rename helper to checkdeps?

fkautz commented on utils/helper/check_deps.go:
```diff
@@ -0,0 +1,99 @@
+	t := dereferenceType(reflect.TypeOf(p))
+	v := dereferenceValue(reflect.ValueOf(p))
+	if t.Kind() != reflect.Struct {
+		return nil, nil, fmt.Errorf("CheckDeps requires a plugin that is a struct,%s is a %s", t, t.Kind())
```
add a space after struct in the comment for readability

fkautz commented on utils/helper/check_deps.go:
```diff
@@ -0,0 +1,99 @@
+		return nil, nil, fmt.Errorf("CheckDeps requires a plugin that is a struct,%s is a %s", t, t.Kind())
+	}
+	sf, ok := t.FieldByName(depsFieldname)
+	if !ok {
```
Is it ever possible that ok would be false at this point? If not, we should probably log this as this indicates an error.

fkautz commented on utils/idempotent/idempotent_impl.go:
```diff
@@ -42,39 +58,61 @@ func (i *Impl) IsIdempotent() bool {
-	if i.refCount < 0 { // i.refCount < 0 means we are terminally closed and no longer Init-able
+	i.Lock()
+	defer i.Unlock()
+	if i.state == CLOSED { // We are closed an no longer initable
```
replace "an" with "and"

edwarnicke commented:
@fkautz in response to:

```
Considering the size of check_deps, would it make sense to make a checkdeps package within helpers? If there is only going to be one helper, perhaps we can rename helper to checkdeps?
```

I'm quite warming to the idea of doing separate package for CheckDeps, InitDeps, CloseDeps, and GetDeps (I split out GitDeps in a subsequent PR to avoid a dependency loop).  Lets hold this patch while I split these up.

edwarnicke commented:
OK, I think this is good to go now.

fkautz commented on utils/helper/deptools/check.go:
```diff
@@ -0,0 +1,68 @@
+			field := v.Interface()
+			zeroValue := reflect.Zero(v.Type())
+			if field == nil || field == zeroValue {
+				return fmt.Errorf("field in Deps struct  named %s is zero valued, and lacks tag `optional:\"true\"`.  Please either set it to a non-zero value or add the `optional:\"true\"` to indicate its acceptable for it to be zero valued", sf.Name)
```
Missed this earlier, struct has two spaces after it instead of one
